plugins/techdocs/src/client.ts

```diff
@@ -49,6 +55,74 @@ export class TechDocsClient implements TechDocsApi {
     this.configApi = options.configApi;
     this.discoveryApi = options.discoveryApi;
     this.fetchApi = options.fetchApi;
+    this.channel.onmessage = event => {
+      const { action, payload } = event.data;
+      if (action === 'REFRESH_COOKIE') {
+        this.locked = payload.locked;
+      }
+    };
+  }
+
+  set expiresAt(value: string) {
+    localStorage.setItem(EXPIRES_AT_STORAGE_KEY, value);
+  }
+
+  get expiresAt(): string | null {
+    return localStorage.getItem(EXPIRES_AT_STORAGE_KEY);
+  }
+
+  private async getCookie(): Promise<{ expiresAt: string }> {
+    const apiOrigin = await this.getApiOrigin();
+    const requestUrl = `${apiOrigin}/cookie`;
+    const response = await this.fetchApi.fetch(`${requestUrl}`, {
+      credentials: 'include',
+    });
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+    return await response.json();
+  }
+
+  private refreshUserCookie(options: { expiresAt: string }) {
+    const { expiresAt } = options;
+    const tenMinutesInMilliseconds = 10 * 60000;
+    const intervalId = setInterval(() => {
+      if (this.locked) return;
+      if (Date.parse(expiresAt) - Date.now() - tenMinutesInMilliseconds > 0)
+        return;
+
+      this.channel.postMessage({
+        action: 'REFRESH_COOKIE',
+        payload: { locked: true },
```

Comment:
This locking mechanism seems a bit dangerous in that if the unlock never happens then it might break other tabs. Perhaps a more safe strategy could be to broadcast a message when refreshing the cookie, and on receipt the other tabs just reset their timeout. By introducing a bit of random jitter to the timeout you can also make sure that tabs generally don't refresh at the same time and there usually is just a single winner.

Reply:
Sounds sensible. And skipping a locking mechanism and accidentally very rarely getting overlapping requests won't be a problem in practice (whichever cookie came last "wins" until the next round, and that's fine)

Reply:
Code updated, thank you for pointing out this 🙌🏻